File: Algorithms/0118_Pascal's_Triangle/Python/Pascal's_Triangle_Solution_1.py

# Space: O(n)
# Time: O(n^2)

class Solution:
    def generate(self, numRows):
        res = []
        prev = []

        for i in range(1, numRows + 1):
            temp = []
            for j in range(i):
                if j == 0 or j >= len(prev):
                    temp.append(1)
                else:
                    temp.append(prev[j] + prev[j - 1])
            res.append(temp)
            prev = temp[:]
        return res

File: Algorithms/0118_Pascal's_Triangle/Java/Pascal'sTriangleSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> res = new ArrayList<>();
        if (numRows == 1) {
            res.add(Arrays.asList(1));
            return res;
        }
        List<Integer> prevRow = null;
        for (int i = 1; i <= numRows; i++) {
            List<Integer> temp = new ArrayList<>();
            for (int j = 0; j < i; j++) {
                if (j == 0 || j == i - 1) {
                    temp.add(1);
                } else {
                    temp.add(prevRow.get(j - 1) + prevRow.get(j));
                }
            }
            prevRow = temp;
            res.add(temp);
        }
        return res;
    }
}




